man/corosync-qnetd-tool.8
.\"/*
.\" * Copyright (C) 2016 Red Hat, Inc.
.\" *
.\" * All rights reserved.
.\" *
.\" * Author: Jan Friesse <jfriesse@redhat.com>
.\" *
.\" * This software licensed under BSD license, the text of which follows:
.\" *
.\" * Redistribution and use in source and binary forms, with or without
.\" * modification, are permitted provided that the following conditions are met:
.\" *
.\" * - Redistributions of source code must retain the above copyright notice,
.\" *   this list of conditions and the following disclaimer.
.\" * - Redistributions in binary form must reproduce the above copyright notice,
.\" *   this list of conditions and the following disclaimer in the documentation
.\" *   and/or other materials provided with the distribution.
.\" * - Neither the name of Red Hat, Inc. nor the names of its
.\" *   contributors may be used to endorse or promote products derived from this
.\" *   software without specific prior written permission.
.\" *
.\" * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
.\" * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
.\" * THE POSSIBILITY OF SUCH DAMAGE.
.\" */
.TH COROSYNC-QNETD-TOOL 8 2016-06-23
.SH NAME
corosync-qnetd-tool \- corosync-qnetd control interface.
.SH SYNOPSIS
.B "corosync-qnetd-tool [-Hhlsv] [-c cluster_name] [-p qnetd_ipc_socket_path]"
.SH DESCRIPTION
.B corosync-qnetd-tool
is a frontend to the internal corosync-qnetd IPC. Its main purpose is to show important
information about the current internal state of
.B corosync-qnetd.
.SH OPTIONS
.TP
.B -H
Properly shutdown the
.B corosync-qnetd
process
.TP
.B -h
Display a short usage text
.TP
.B -l
List all clients connected to the
.B corosync-qnetd
process. The output is described in its own section below.
.TP
.B -s
Display status of the
.B corosync-qnetd
process.
.TP
.B -v
Display more verbose output for options
.B -l
and
.B -s
.TP
.B -c
Used only with the
.B -l
option. By default, information about all clients from all clusters is displayed, with
this option it's possible to filter information from a single cluster given the
.I cluster_name.
.TP
.B -p
Path to the
.B corosync-qnetd
communication socket.

.SH LIST COMMAND OUTPUT
.nf
Cluster "Cluster":
    Algorithm:          Fifty-Fifty split
    Tie-breaker:        Node with lowest node ID
    Node ID 1:
        Client address:         ::ffff:127.0.0.1:52000
        HB interval:            8000ms
        Configured node list:   1, 2
        Ring ID:                1.a00000000021b40
        Membership node list:   1, 2
        TLS active:             Yes (client certificate verified)
        Vote:                   No change (ACK)
 ...
.fi

The output contains a list of clusters. Each cluster has the cluster common options
.I Algorithm
and
.I Tie-breaker
as configured in the corosync.conf file. Information about nodes follows.
.I Client address
is the IP address and port of the client.
.I HB interval
is the heartbeat interval between
.B corosync-qnetd
and
.B corosync-qdevice
client. This option can be configured in corosync.conf.
.I Configured node list
is the list of nodes configured in corosync.conf.
.I Ring ID
and
.I Membership node list
are self-explanatory.
.I TLS active
describes if an encrypted transport is used between server and client.
.I Vote
is last vote sent to
.B corosync-qdevice
client. The last ACK/NACK vote (if it exists) is in parentheses.
.SH SEE ALSO
.BR corosync-qnetd (8)
.BR corosync-qdevice (8)
.SH AUTHOR
Jan Friesse
.PP


CHANGELOG
------------------------------------------------------------------------
r3035 | sdake | 2010-08-31 12:04:23 -0700 (Tue, 31 Aug 2010) | 6 lines

Merge trunk revision 3034:
r3034 | sdake | 2010-08-31 12:02:19 -0700 (Tue, 31 Aug 2010) | 2 lines

Bump version to 1.2.8.


------------------------------------------------------------------------
r3033 | sdake | 2010-08-31 12:01:51 -0700 (Tue, 31 Aug 2010) | 7 lines

Merge trunk revision 3030:
r3030 | sdake | 2010-08-31 11:55:27 -0700 (Tue, 31 Aug 2010) | 3 lines

Build properly with different versions of libc by including all posix APIs
in header definitions.


------------------------------------------------------------------------
r3032 | sdake | 2010-08-31 12:01:27 -0700 (Tue, 31 Aug 2010) | 6 lines

Merge trunk revision 3029:
r3029 | asalkeld | 2010-08-24 18:13:14 -0700 (Tue, 24 Aug 2010) | 2 lines

IPC: return CS_ERR_NO_RESOURCES to library when low on fds.


------------------------------------------------------------------------
r3031 | sdake | 2010-08-31 12:00:52 -0700 (Tue, 31 Aug 2010) | 6 lines

Merge trunk revision 3028:
r3028 | asalkeld | 2010-08-24 18:07:37 -0700 (Tue, 24 Aug 2010) | 2 lines
    
POLL: gracefully handle running out of file descriptors.


------------------------------------------------------------------------
r3027 | sdake | 2010-08-24 13:28:46 -0700 (Tue, 24 Aug 2010) | 5 lines

Merge trunk revision 3024:
r3024 | sdake | 2010-08-24 11:45:43 -0700 (Tue, 24 Aug 2010) | 2 lines

Remove checking of sub parameters in service.d files.

------------------------------------------------------------------------
r3026 | sdake | 2010-08-24 13:09:54 -0700 (Tue, 24 Aug 2010) | 9 lines

Merge trunk revision 3011:
r3011 | honzaf | 2010-08-02 05:38:19 -0700 (Mon, 02 Aug 2010) | 5 lines
    
Check that Corosync was not runned by cman
    
Patch fixes init script so it's tested, if corosync was runned
by cman or not. If so, it refuses to stop Corosync.


------------------------------------------------------------------------
r3025 | sdake | 2010-08-24 13:07:05 -0700 (Tue, 24 Aug 2010) | 11 lines

Merge trunk revision 3010:
r3010 | honzaf | 2010-08-02 05:36:20 -0700 (Mon, 02 Aug 2010) | 7 lines

Allow running only one instance of Corosync

Patch makes Corosync more compliant with common practices
for writing daemon. It creates pid file (usually
/var/run/corosync.pid) and flocks it. So only one instance
of Corosync can be executed now.


------------------------------------------------------------------------
r3023 | sdake | 2010-08-17 11:11:25 -0700 (Tue, 17 Aug 2010) | 6 lines

Merge trunk revision 3022:
r3022 | sdake | 2010-08-17 11:08:13 -0700 (Tue, 17 Aug 2010) | 2 lines

Properly detect shutdown of corosync process


------------------------------------------------------------------------
r3013 | sdake | 2010-08-03 10:36:37 -0700 (Tue, 03 Aug 2010) | 6 lines

Merge trunk revision 3012:
r3012 | sdake | 2010-08-03 10:31:33 -0700 (Tue, 03 Aug 2010) | 2 lines

Remove cancel token retransmit timeout.


------------------------------------------------------------------------
r3008 | sdake | 2010-07-27 14:37:46 -0700 (Tue, 27 Jul 2010) | 6 lines

Merge trunk revision 3007:
r3007 | sdake | 2010-07-27 14:32:39 -0700 (Tue, 27 Jul 2010) | 2 lines

Change trunk version to 1.2.7.


------------------------------------------------------------------------
r3006 | sdake | 2010-07-27 12:08:29 -0700 (Tue, 27 Jul 2010) | 7 lines

Merge trunk revision 3005:
r3005 | sdake | 2010-07-27 12:00:37 -0700 (Tue, 27 Jul 2010) | 3 lines

Remove consensus check for two node cluster cases which can have smaller
consensus values.  Document in man page the behavior of consensus.


------------------------------------------------------------------------
r3004 | asalkeld | 2010-07-21 16:16:41 -0700 (Wed, 21 Jul 2010) | 5 lines

Merge trunk revision 3002:
r3002 | sdake | 2010-07-22 06:48:40 +1000 (Thu, 22 Jul 2010) | 2 lines
Fix merge error with revision 3001.


------------------------------------------------------------------------
r3003 | asalkeld | 2010-07-21 16:13:42 -0700 (Wed, 21 Jul 2010) | 6 lines

Merge trunk revision 3001:
r3001 | sdake | 2010-07-22 03:03:36 +1000 (Thu, 22 Jul 2010) | 3 lines
Fix problem where flow control could lock up ipc under very heavy load in very
rare circumstances.


------------------------------------------------------------------------
r3000 | sdake | 2010-07-19 13:20:25 -0700 (Mon, 19 Jul 2010) | 12 lines

Merge trunk revision 2995:
r2995 | asalkeld | 2010-07-16 21:59:40 -0700 (Fri, 16 Jul 2010) | 9 lines

SYNC: always call sync_aborted() in sync_confchg_fn().

1) sync_callbacks.sync_abort can be null.
2) sync_processing is set to 0 after syncv1 is done.
   Then syncv2 processing is down. If we get a config change
   after syncv1 is down, but before syncv2 is done then it won't
   get aborted.


------------------------------------------------------------------------
r2999 | sdake | 2010-07-19 13:19:30 -0700 (Mon, 19 Jul 2010) | 6 lines

Merge trunk revision 2991:
r2991 | asalkeld | 2010-07-15 19:08:54 -0700 (Thu, 15 Jul 2010) | 2 lines

 SYNCV2: reset the my_memb_determine_ring_id in sync_v2_memb_list_abort()


------------------------------------------------------------------------
r2998 | sdake | 2010-07-19 08:29:30 -0700 (Mon, 19 Jul 2010) | 7 lines

Merge trunk revision 2997:
r2997 | fabbione | 2010-07-18 23:36:48 -0700 (Sun, 18 Jul 2010) | 4 lines

Fix logging_daemon config parser code.

Resolves: rhbz#615203

------------------------------------------------------------------------
r2989 | sdake | 2010-07-14 11:36:53 -0700 (Wed, 14 Jul 2010) | 8 lines

Merge trunk revision 2988:
r2988 | sdake | 2010-07-14 11:35:36 -0700 (Wed, 14 Jul 2010) | 4 lines

Remove reset of token timeout on retransmitted token reception.  The timer
should only be reset when a real token is received or membership protocol
could run into problems with certain timing parameters.



qdevices/corosync-qnetd-tool.c
/*
 * Copyright (c) 2015-2016 Red Hat, Inc.
 *
 * All rights reserved.
 *
 * Author: Jan Friesse (jfriesse@redhat.com)
 *
 * This software licensed under BSD license, the text of which follows:
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * - Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 * - Neither the name of the Red Hat, Inc. nor the names of its
 *   contributors may be used to endorse or promote products derived from this
 *   software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 * THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <err.h>
#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "qnet-config.h"

#include "dynar.h"
#include "dynar-str.h"
#include "utils.h"
#include "unix-socket.h"

#define IPC_READ_BUF_SIZE	512

enum qnetd_tool_operation {
	QNETD_TOOL_OPERATION_NONE,
	QNETD_TOOL_OPERATION_SHUTDOWN,
	QNETD_TOOL_OPERATION_STATUS,
	QNETD_TOOL_OPERATION_LIST,
};

enum qnetd_tool_exit_code {
	QNETD_TOOL_EXIT_CODE_NO_ERROR = 0,
	QNETD_TOOL_EXIT_CODE_USAGE = 1,
	QNETD_TOOL_EXIT_CODE_INTERNAL_ERROR = 2,
	QNETD_TOOL_EXIT_CODE_SOCKET_CONNECT = 3,
	QNETD_TOOL_EXIT_CODE_QNETD_RETURNED_ERROR = 4,
};

static void
usage(void)
{

	printf("usage: %s [-Hhlsv] [-c cluster_name] [-p qnetd_ipc_socket_path]\n",
	    QNETD_TOOL_PROGRAM_NAME);
}

static void
cli_parse(int argc, char * const argv[], enum qnetd_tool_operation *operation,
    int *verbose, char **cluster_name, char **socket_path)
{
	int ch;

	*operation = QNETD_TOOL_OPERATION_NONE;
	*verbose = 0;
	*cluster_name = NULL;
	*socket_path = strdup(QNETD_DEFAULT_LOCAL_SOCKET_FILE);

	if (*socket_path == NULL) {
		errx(QNETD_TOOL_EXIT_CODE_INTERNAL_ERROR,
		    "Can't alloc memory for socket path string");
	}

	while ((ch = getopt(argc, argv, "Hhlsvc:p:")) != -1) {
		switch (ch) {
		case 'H':
			*operation = QNETD_TOOL_OPERATION_SHUTDOWN;
			break;
		case 'l':
			*operation = QNETD_TOOL_OPERATION_LIST;
			break;
		case 's':
			*operation = QNETD_TOOL_OPERATION_STATUS;
			break;
		case 'v':
			*verbose = 1;
			break;
		case 'c':
			free(*cluster_name);
			*cluster_name = strdup(optarg);
			if (*cluster_name == NULL) {
				errx(QNETD_TOOL_EXIT_CODE_INTERNAL_ERROR,
				    "Can't alloc memory for cluster name string");
			}
			break;
		case 'p':
			free(*socket_path);
			*socket_path = strdup(optarg);
			if (*socket_path == NULL) {
				errx(QNETD_TOOL_EXIT_CODE_INTERNAL_ERROR,
				    "Can't alloc memory for socket path string");
			}
			break;
		case 'h':
		case '?':
			usage();
			exit(QNETD_TOOL_EXIT_CODE_USAGE);
			break;
		}
	}

	if (*operation == QNETD_TOOL_OPERATION_NONE) {
		usage();
		exit(QNETD_TOOL_EXIT_CODE_USAGE);
	}
}

static int
store_command(struct dynar *str, enum qnetd_tool_operation operation, int verbose,
    const char *cluster_name)
{
	const char *nline = "\n\0";
	const int nline_len = 2;

	switch (operation) {
	case QNETD_TOOL_OPERATION_NONE:
		errx(QNETD_TOOL_EXIT_CODE_INTERNAL_ERROR, "Unhandled operation none");
		break;
	case QNETD_TOOL_OPERATION_SHUTDOWN:
		if (dynar_str_cat(str, "shutdown ") != 0) {
			return (-1);
		}
		break;
	case QNETD_TOOL_OPERATION_STATUS:
		if (dynar_str_cat(str, "status ") != 0) {
			return (-1);
		}
		break;
	case QNETD_TOOL_OPERATION_LIST:
		if (dynar_str_cat(str, "list ") != 0) {
			return (-1);
		}
		break;
	}

	if (verbose) {
		if (dynar_str_cat(str, "verbose ") != 0) {
			return (-1);
		}
	}

	if (cluster_name != NULL) {
		if (dynar_str_cat(str, "cluster ") != 0 ||
		    dynar_str_quote_cat(str, cluster_name) != 0 ||
		    dynar_str_cat(str, " ") != 0) {
			return (-1);
		}
	}

	if (dynar_cat(str, nline, nline_len) != 0) {
		return (-1);
	}

	return (0);
}

/*
 * -1 - Internal error (can't alloc memory)
 *  0 - No error
 *  1 - IPC returned error
 *  2 - Unknown status line
 */
static int
read_ipc_reply(FILE *f)
{
	struct dynar read_str;
	int ch;
	int status_readed;
	int res;
	static const char *ok_str = "OK";
	static const char *err_str = "Error";
	int err_set;
	char c;

	dynar_init(&read_str, IPC_READ_BUF_SIZE);

	status_readed = 0;
	err_set = 0;
	res = 0;

	while ((ch = fgetc(f)) != EOF) {
		if (status_readed) {
			putc(ch, (err_set ? stderr : stdout));
		} else {
			if (ch == '\r') {
			} else if (ch == '\n') {
				status_readed = 1;

				c = '\0';
				if (dynar_cat(&read_str, &c, sizeof(c)) != 0) {
					res = -1;
					goto exit_destroy;
				}

				if (strcasecmp(dynar_data(&read_str), ok_str) == 0) {
				} else if (strcasecmp(dynar_data(&read_str), err_str) == 0) {
					err_set = 1;
					res = 1;
					fprintf(stderr, "Error: ");
				} else {
					res = 2;
					goto exit_destroy;
				}
			} else {
				c = ch;
				if (dynar_cat(&read_str, &c, sizeof(c)) != 0) {
					res = -1;
					goto exit_destroy;
				}
			}
		}
	}

exit_destroy:
	dynar_destroy(&read_str);

	return (res);
}

int
main(int argc, char * const argv[])
{
	enum qnetd_tool_operation operation;
	int verbose;
	char *cluster_name;
	char *socket_path;
	int sock_fd;
	FILE *sock;
	struct dynar send_str;
	int res;
	int exit_code;

	exit_code = QNETD_TOOL_EXIT_CODE_NO_ERROR;

	cli_parse(argc, argv, &operation, &verbose, &cluster_name, &socket_path);

	dynar_init(&send_str, QNETD_DEFAULT_IPC_MAX_RECEIVE_SIZE);

	sock_fd = unix_socket_client_create(socket_path, 0);
	if (sock_fd == -1) {
		err(QNETD_TOOL_EXIT_CODE_SOCKET_CONNECT,
		    "Can't connect to qnetd socket (is QNetd running?)");
	}

	sock = fdopen(sock_fd, "w+t");
	if (sock == NULL) {
		err(QNETD_TOOL_EXIT_CODE_INTERNAL_ERROR, "Can't open QNetd socket fd");
	}

	if (store_command(&send_str, operation, verbose, cluster_name) != 0) {
		errx(QNETD_TOOL_EXIT_CODE_INTERNAL_ERROR, "Can't store command");
	}

	if (fprintf(sock, "%s", dynar_data(&send_str)) != strlen(dynar_data(&send_str)) ||
	    fflush(sock) != 0) {
		errx(QNETD_TOOL_EXIT_CODE_INTERNAL_ERROR, "Can't send command");
	}

	res = read_ipc_reply(sock);
	switch (res) {
	case -1:
		errx(QNETD_TOOL_EXIT_CODE_INTERNAL_ERROR, "Internal error during IPC status line read");
		break;
	case 0:
		break;
	case 1:
		exit_code = QNETD_TOOL_EXIT_CODE_QNETD_RETURNED_ERROR;
		break;
	case 2:
		errx(QNETD_TOOL_EXIT_CODE_SOCKET_CONNECT, "Unknown status line returned by IPC server");
		break;
	}

	if (fclose(sock) != 0) {
		warn("Can't close QNetd socket");
	}

	free(cluster_name);
	free(socket_path);
	dynar_destroy(&send_str);

	return (exit_code);
}
